### ISSUE TITLE
WVJB for Android

### ISSUE THREAD

**Issue by marcuswestin** [marcuswestin/WebViewJavascriptBridge#103]
Review and consider merging in Android work done by @fangj in https://github.com/fangj/WebViewJavascriptBridge.

**Comment by tonycn** [comment]
Exciting  features!

**Comment by yangfeiyu** [comment]
Excitied!

**Comment by vary** [comment]
nice

**Comment by zhiyongXu** [comment]
此邮件来自易信
网易邮箱推出官方手机客户端--邮箱大师 马上体验


 在2016年1月12日 16:59:10, lishaoping <notifications@github.com> 写道：



nice

**Comment by pjq** [comment]
waiting for the Android compatible implementation.

**Comment by THEONE10211024** [comment]
waiting！！！！！

**Comment by marcuswestin** [comment]
If you build it and PR, I'll totally merge in :)

**Comment by ssomai** [comment]
waiting too~

**Comment by jojonarte** [comment]
Any news on this feature?

**Comment by marcuswestin** [comment]
There's https://github.com/marcuswestin/WebViewJavascriptBridge/pull/217, but it doesn't follow the API of the existing ObjC WebViewJavascriptBridge, and isn't very clean code. I'm still willing to accept a proper implementation.

**Comment by gedw99** [comment]
@marcuswestin 
I guess the long time between https://github.com/fangj/WebViewJavascriptBridge submitted the PR and you responding caused this to collapse.

Anyway, it seems a shame to not get this done as i am sure its really useful for many people.

i had a look at the code and also your request to a code cleanup.
I noticed that you also offered to document where you see things need to be changed to make the code clean. Would you will willing to do that still ?
I think its highly likely that me and hopefully others can do that work.

**Comment by marcuswestin** [comment]
@gedw99 Yeah, having a kid, early stage startups and solo-maintaining multiple open source projects don't go together very well :)

I don't think fangj ever did a PR though, did he? The only Android PR I recall is this #217, which is a good start but not as good as fangj's I believe. Only looked closely enough to know it wasn't ready for primetime.

I would love that! If you take ownership I will make sure to be available for any discussions and questions with minimal turnaround time. How about I start with a rough sketch of what I think we'd need, and you start from there?

Let me know what you think!

## Android WVJB - rough sketch

I think the guiding principle of an Android version should be to mirror the ObjC version, except when there are strong reasons not to (in which case we should consider reshaping the ObjC version to mirror the Android one). The motivation is two-fold: First, understanding one version should make it very simple to understand and navigate the other; and second, when we add features to one version it should be simple to also add it to the other. I believe there are five major areas where we need parity and similarity: API design, native code, javascript code, automated testing coverage, and documentation.

### API design mirroring

The Android & ObjC APIs should basically be identical, excepting language-idiomatic differences. I haven't done much Android work for a while so I'm probably not the best person to say, but I'm guessing it would be something similar to this:

```java
WebViewJavascriptBridge* bridge = WebViewJavascriptBridge.bridgeForWebView(webView);
bridge.registerHandler("getScreenHeight", (message, responseCallback) -> {
	responseCallback(message);
});
```

### Native code mirroring:

This one is a bit of a stretch and may not make sense at the end of the day, but my guess is that the most appropriate implementation would look very similar in layout and structure to the ObjC WVJB base:
- https://github.com/marcuswestin/WebViewJavascriptBridge/blob/master/WebViewJavascriptBridge/WebViewJavascriptBridgeBase.m

### Javascript code mirroring:

The ObjC and Android code bases should use the same javascript code except if there are strong reasons not to. The two javascript pieces are:

- bootstrap/setup snippet: https://github.com/marcuswestin/WebViewJavascriptBridge/blob/master/Example%20Apps/ExampleApp.html#L17
- injected javascript API: https://github.com/marcuswestin/WebViewJavascriptBridge/blob/master/WebViewJavascriptBridge/WebViewJavascriptBridge_JS.m

I imagine it would make sense for the injected javascript API to be compiled from a shared source file at build time, but that's a small detail.

### Automated testing mirroring

The Android version's automated testing would need to cover the same ground as the ObjC version, and be passing on travis-ci:

- https://github.com/marcuswestin/WebViewJavascriptBridge/blob/master/Tests/WebViewJavascriptBridgeTests/BridgeTests.m
- https://travis-ci.org/marcuswestin/WebViewJavascriptBridge

### Documentation mirroring

There are a few sections in Readme.md that would need Android entries:

- Installation instructions: https://github.com/marcuswestin/webviewjavascriptbridge#installation-ios--osx
- Small usage guide: https://github.com/marcuswestin/webviewjavascriptbridge#usage
- Java API: https://github.com/marcuswestin/webviewjavascriptbridge#objc-api
- Example apps: https://github.com/marcuswestin/webviewjavascriptbridge#examples (always the easiest way to grok a new library)

I imagine the installation instructions would cover whatever the commonly used IDEs are, though they could of course be fleshed out with time.

Considering the Readme.md file would start getting a bit unwieldy at that point it may make more sense to have Readme.md just link to e.g Readme-apple.md, Readme-android.md and Readme-javascript.md.

## Android version support

Finally, whoever takes on ownership of developing the Android version of WVJB would also take on support and maintenance, and/or involve other people who take it on. Roughly, that breaks down to responding to any android issues on GH, and manage, clean up and merge any android PRs.

Whoever takes this on would naturally get full r/w access to the repo!

**Comment by marcuswestin** [comment]
Congrats! That's super exciting :)

Oh well, if/when you change your mind there's at least a clear direction to start in.

Cheers!

**Comment by sujiewen** [comment]
Cattle people, how about Android?

**Comment by marcuswestin** [comment]
Related: #217

**Comment by sujiewen** [comment]
That's too bad

**Comment by wendux** [comment]
Another implementation for Android is : 
 https://github.com/wendux/WebViewJavascriptBridge
It is a mirror of  [marcuswestin/WebViewJavascriptBridge](https://github.com/marcuswestin/WebViewJavascriptBridge), there are some similarities between the two project such as API design, native code, and the example demo. For the Javascript code, it is exactly the same.  More details please refer to #329
